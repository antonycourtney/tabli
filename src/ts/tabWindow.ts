/*
 * Representation of tabbed windows using Immutable.js
 */
import { log } from './globals';
import * as Immutable from 'immutable';
import { escapeTableCell } from './utils';

export interface SavedTabStateProps {
    bookmarkId: string;
    bookmarkIndex: number;
    title: string;
    url: string;
}
const defaultSavedTabStateProps: SavedTabStateProps = {
    bookmarkId: '',
    bookmarkIndex: 0, // position in bookmark folder
    title: '',
    url: '',
};

/*
 * Tab state that is persisted as a bookmark
 */
export class SavedTabState extends Immutable.Record(defaultSavedTabStateProps) {
    static fromJS(js: any): SavedTabState | null {
        if (js === null) {
            return null;
        }
        return new SavedTabState(js);
    }
}

export interface OpenTabStateProps {
    url: string;
    openTabId: number;
    active: boolean;
    openTabIndex: number;
    favIconUrl: string;
    title: string;
    audible: boolean;
    muted: boolean;
    pinned: boolean;
    isSuspended: boolean;
    openerTabId: number | undefined;
    openerUrl: string | undefined; // url of openerTabId at time tab created
}

const defaultOpenTabStateProps: OpenTabStateProps = {
    url: '',
    openTabId: -1,
    active: false,
    openTabIndex: 0, // index of open tab in its window
    favIconUrl: '',
    title: '',
    audible: false,
    muted: false,
    pinned: false,
    isSuspended: false,
    openerTabId: undefined,
    openerUrl: undefined,
};

/*
 * Tab state associated with an open browser tab
 */
export class OpenTabState extends Immutable.Record(defaultOpenTabStateProps) {
    static fromJS(js: any): OpenTabState | null {
        if (js === null) {
            return null;
        }
        return new OpenTabState(js);
    }
}

interface TabItemProps {
    saved: boolean;
    savedState: SavedTabState | null;
    open: boolean;
    openState: OpenTabState | null;
}

const defaultTabItemProps: TabItemProps = {
    saved: false,
    savedState: null, // SavedTabState iff saved

    open: false, // Note: Saved tabs may be closed even when containing window is open
    openState: null, // OpenTabState iff open
};

// Use a counter to generate unique keys for tab items on-demand...
// lets
let tabItemKeyCounter = 500;

const genTabItemKey = () => '_tabItem-' + tabItemKeyCounter++;

/*
 * An item in a tabbed window.
 *
 * May be associated with an open tab, a bookmark, or both
 */
export class TabItem extends Immutable.Record(defaultTabItemProps) {
    private _id: string | undefined;

    get title(): string {
        /*
        hold-over from exploring showing saved titles instead of current title:
        if (this.saved && this.open) {
            if (this.savedState!.title !== this.openState!.title) {
                log.debug(
                    'tabItem saved title: ',
                    this.savedState!.title,
                    '|',
                    this.openState!.title
                );
            }
        }
        */
        if (this.open && this.openState) {
            return this.openState.title;
        }

        return this.savedState ? this.savedState.title : '';
    }

    // get a unique key for this TabItem
    // useful as key in React arrays
    // Note: not unique across persisted sessions / Chrome restart
    // We briefly tried to use open tab ids or bookmarks ids, but this gets horribly confusing
    // and risks collisions when, for example, we persist the last opened state of a tab and
    // re-hydrate it.
    get key(): string {
        if (this._id === undefined) {
            this._id = genTabItemKey();
        }
        return this._id;
    }

    // just for debugging:
    get rawKey(): string | undefined {
        return this._id;
    }

    get url(): string {
        if (this.open && this.openState) {
            return this.openState.url;
        }

        return this.savedState ? this.savedState.url : '';
    }

    get pinned(): boolean {
        if (this.open && this.openState) {
            return this.openState.pinned;
        }
        return false;
    }

    get active(): boolean {
        if (this.open && this.openState) {
            return this.openState.active;
        }
        return false;
    }

    // safe accessor for savedState:
    get safeSavedState(): SavedTabState {
        if (this.saved && this.savedState) {
            return this.savedState;
        }
        throw new Error('Unexpected access of savedState on unsaved tab');
    }

    // safe accessor for openState:
    get safeOpenState(): OpenTabState {
        if (this.open && this.openState) {
            return this.openState;
        }
        throw new Error('Unexpected access of openState on non-open tab');
    }

    static fromJS(js: any): TabItem {
        const savedState = SavedTabState.fromJS(js.savedState);
        const openState = OpenTabState.fromJS(js.openState);

        return new TabItem({
            saved: js.saved,
            savedState,
            open: js.open,
            openState,
        });
    }
}

/*
 * A TabWindow
 *
 * Tab windows have a title and a set of tab items.
 *
 * A TabWindow has 4 possible states:
 *   (open,!saved)   - An open Chrome window that has not had its tabs saved
 *   (open,saved)    - An open Chrome window that has also had its tabs saved (as bookmarks)
 *   (!open,saved,!snapshot)   - A previously saved window that is not currently
 *                           open and has no snapshot. tabItems will consist solely
 *                           of saved tabs (persisted as boookmarks).
 *
 *   (!open,saved,snapshot) - A previously saved window that is not currently open but
 *                            where tabItems contains the tab state the last time the
 *                            window was open
 */
interface TabWindowProps {
    saved: boolean;
    savedTitle: string;
    savedFolderId: string;

    open: boolean;
    openWindowId: number;
    windowType: string;
    width: number;
    height: number;
    tabItems: Immutable.List<TabItem>;
    snapshot: boolean;
    chromeSessionId: string | null;
    expanded: boolean | null;
}

const defaultTabWindowProps: TabWindowProps = {
    saved: false,
    savedTitle: '',
    savedFolderId: '',

    open: false,
    openWindowId: -1,
    windowType: '',
    width: 0,
    height: 0,

    tabItems: Immutable.List(), // <TabItem>

    snapshot: false, // Set if tabItems contains snapshot of last open state
    chromeSessionId: null, // Chrome session id for restore (if found)

    // This is view state, so technically doesn't belong here, but we only have
    // one window component per window right now, we want to be able to toggle
    // this state via a keyboard handler much higher in the hierarchy,
    // and cost to factor out view state would be high.
    expanded: null, // tri-state: null, true or false
};

export class TabWindow extends Immutable.Record(defaultTabWindowProps) {
    private _title: string | undefined;
    private _key: string | undefined;

    get title(): string {
        if (this._title === undefined) {
            this._title = this.computeTitle();
        }

        return this._title;
    }

    computeTitle(): string {
        if (this.saved) {
            return this.savedTitle;
        }

        const activeTab = this.tabItems.find(
            (t) => t.open && t.openState!.active,
        );

        if (!activeTab) {
            // shouldn't happen!
            log.debug(
                'TabWindow.get title(): No active tab found: ',
                this.toJS(),
            );

            var openTabItem = this.tabItems.find((t) => t.open);
            if (!openTabItem) {
                return '';
            }
            return openTabItem.title;
        }

        return activeTab.title;
    }

    updateSavedTitle(title: string): TabWindow {
        delete this._title;
        return this.set('savedTitle', title);
    }

    // get a unique key for this window
    // useful as key in React arrays
    // Should be adequate as a key for React, but does not uniquely
    // determine the immutable record state.
    get key(): string {
        if (this._key === undefined) {
            if (this.saved) {
                this._key = '_saved' + this.savedFolderId;
            } else {
                this._key = '_open' + this.openWindowId;
            }
        }
        return this._key;
    }

    get openTabCount(): number {
        return this.tabItems.count((ti) => ti.open);
    }

    /*
     * Returns [index,TabItem] pair if window contains chrome tab id or else null
     */
    findChromeTabId(tabId: number): [number, TabItem] | null {
        const ret = this.tabItems.findEntry(
            (ti) =>
                !!ti.open &&
                ti.openState !== null &&
                ti.openState.openTabId === tabId,
        );
        return ret ? ret : null;
    }
    /*
     * Returns [index,TabItem] pair if window contains chrome bookmark id or else null
     */
    findChromeBookmarkId(bookmarkId: string): [number, TabItem] | null {
        const ret = this.tabItems.findEntry(
            (ti) =>
                !!ti.saved &&
                ti.savedState !== null &&
                ti.savedState.bookmarkId === bookmarkId,
        );
        return ret ? ret : null;
    }

    findTabByKey(key: string): [number, TabItem] | null {
        const ret = this.tabItems.findEntry((ti) => ti.key === key);
        return ret ? ret : null;
    }

    getActiveTabId(): number | null {
        const activeTab = this.tabItems.find(
            (t) => t.open && t.openState!.active,
        );
        const tabId = activeTab ? activeTab.openState!.openTabId : null;
        return tabId;
    }

    /*
     * set tabItems.
     * Used to sort, but openTabIndex from Chrome isn't maintained by tab updates
     * so we reverted that.
     * Then we briefly used .cacheResult() when tabItems was (supposedly) a Seq,
     * but it turned out that certain code paths (such as inserting a tab) would
     * result in nextItems being a List.
     * Seems to make the most sense to just make tabItems a List not a Seq
     */
    setTabItems(nextItems: Immutable.List<TabItem>): TabWindow {
        /* HACK: debugging only check; get rid of this! */
        if (!nextItems) {
            log.error('setTabItems: bad nextItems: ', nextItems);
        }
        return this.set('tabItems', nextItems);
    }

    /*
     * get index of an item in this TabWindow
     */
    indexOf(target: TabItem): number | null {
        return this.tabItems.indexOf(target);
    }

    exportStr(): string {
        const fmtTabItem = (ti: TabItem) => {
            const urls = ti.url != null ? ti.url : '';
            const ret = escapeTableCell(ti.title) + ' | ' + urls + '\n';
            return ret;
        };
        const titleStr = '### ' + this.title;
        const headerStr = `
Title                                  | URL
---------------------------------------|-----------
`;
        const s0 = titleStr + '\n' + headerStr;
        const s = this.tabItems.reduce((rs, ti) => rs + fmtTabItem(ti), s0);
        return s;
    }

    // combine local expanded state with global expanded state to determine if expanded:
    isExpanded(expandAll: boolean): boolean {
        if (this.expanded === null) {
            return expandAll && this.open;
        }
        return this.expanded;
    }

    static fromJS(js: any): TabWindow {
        const tabItems = Immutable.List<TabItem>(
            js.tabItems.map(TabItem.fromJS),
        );
        return new TabWindow({
            saved: js.saved,
            savedTitle: js.savedTitle,
            savedFolderId: js.savedFolderId,
            open: js.open,
            openWindowId: js.openWindowId,
            windowType: js.windowType,
            width: js.width,
            height: js.height,
            tabItems,
            snapshot: js.snapshot,
            chromeSessionId: js.chromeSessionId,
            expanded: js.expanded,
        });
    }
}
