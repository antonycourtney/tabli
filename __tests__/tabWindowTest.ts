import * as TabWindow from '../src/ts/tabWindow';
import * as tabWindowUtils from '../src/ts/tabWindowUtils';
import difflet from 'difflet';
import * as rawTestData from '../test-data/testData';
import log from 'loglevel';
import { enablePatches } from 'immer';

const testData = rawTestData as any;

function dumpDiffs(objA, objB) {
    var s = difflet({ indent: 2, comment: true }).compare(objA, objB);
    console.log('diffs:');
    console.log(s);
}

beforeAll(() => {
    enablePatches();
});

test('basic test', () => {});

test('makeFolderTabWindow', () => {
    const tabWindow = tabWindowUtils.makeFolderTabWindow(
        testData.d3BookmarkFolder,
    );
    const tabWindowJS = JSON.parse(JSON.stringify(tabWindow.toJS()));
    /*
      console.log("makeFolderTabWindow returned:")
      console.log(">>>>>>>")
      console.log(JSON.stringify(tabWindowJS,null,2))
      console.log(">>>>>>>")

      console.log('diffs between tabWindowJS and expected:')
      dumpDiffs(tabWindowJS, testData.d3InitialExpectedTabWindow)
  */

    expect(tabWindowJS).toMatchSnapshot();

    expect(tabWindow.title).toBe('d3 docs');
});

test('chromeTabWindow', () => {
    const baseTabWindow = tabWindowUtils.makeChromeTabWindow(
        testData.chromeWindowSnap,
    );

    const tabWindowJS = JSON.parse(JSON.stringify(baseTabWindow.toJS()));

    /*
      console.log("makeChromeTabWindow returned: ")
      console.log(">>>>>")
      console.log(JSON.stringify(tabWindowJS,null,2))
      console.log(">>>>>")
  */

    // dumpDiffs(tabWindowJS, expectedTabWindow);

    expect(tabWindowJS).toMatchSnapshot();

    // Check that title matches active tab:
    const baseTitle = baseTabWindow.title;
    expect(baseTitle).toBe('Declare Permissions - Google Chrome');

    // Now let's take state after closing the active tab that gave us our title:
    const updTabWindow = tabWindowUtils.updateWindow(
        baseTabWindow,
        testData.chromeWindowSnap2,
    );
    const updTabWindowJS = JSON.parse(JSON.stringify(updTabWindow.toJS()));

    // console.log("Updated tab window: ")
    // console.log(JSON.stringify(updTabWindowJS,null,2))

    const updTitle = updTabWindow.title;
    expect(updTitle).toBe('Fluxxor - Home');
});

/*
 * make sure we still get a deterministic title even when no tab active
 */
test('noActiveTabTitle', () => {
    const baseTabWindow = tabWindowUtils.makeChromeTabWindow(
        testData.chromeWindowSnap3,
    );

    const tabWindowJS = JSON.parse(JSON.stringify(baseTabWindow.toJS()));

    const baseTitle = baseTabWindow.title;

    expect(baseTitle).toBe(testData.chromeWindowSnap3.tabs[0].title);
});

test('missingSourceFieldsTests', () => {
    const bmFolder0 = {
        children: [
            {
                dateAdded: 1395768341441,
                id: '432',
                index: 0,
                parentId: '431',

                // title deliberately omitted
                url: 'https://github.com/mbostock/d3/wiki/API-Reference',
            },
        ],
        dateAdded: 1395768341427,
        dateGroupModified: 1430260300118,
        id: '431',
        index: 4,
        parentId: '377',
        title: 'd3 docs',
    };

    const tabWindow = tabWindowUtils.makeFolderTabWindow(
        bmFolder0 as any,
        false,
    );
    const tabWindowJS = JSON.parse(JSON.stringify(tabWindow.toJS()));

    // console.log("makeFolderTabWindow returned:")
    // console.log(JSON.stringify(tabWindowJS,null,2))

    const bmTabItem = tabWindow.tabItems[0];
    const bmTabTitle = bmTabItem.title;
    expect(bmTabTitle).toBe(bmFolder0.children[0].url);

    // window is still folder title:
    const windowTitle = tabWindow.title;

    // Let's use URL for title of untitled tabs:
    expect(windowTitle).toBe(bmFolder0.title);

    // TODO: Now omit folder title...
    const bmFolder1 = {
        children: [
            {
                dateAdded: 1395768341441,
                id: '432',
                index: 0,
                parentId: '431',
                title: 'd3 API Reference',
                url: 'https://github.com/mbostock/d3/wiki/API-Reference',
            },
        ],
        dateAdded: 1395768341427,
        dateGroupModified: 1430260300118,
        id: '431',
        index: 4,
        parentId: '377',

        // title deliberately omitted
    };

    const tabWindow1 = tabWindowUtils.makeFolderTabWindow(
        bmFolder1 as any,
        false,
    );
    const tabWindow1JS = JSON.parse(JSON.stringify(tabWindow1.toJS()));

    const tabWindow1Title = tabWindow1.title;

    // revert to title of first tab:
    // expect(tabWindow1Title).toBe(bmFolder1.children[0].title);

    const chromeWindowSnap4 = {
        alwaysOnTop: false,
        focused: false,
        height: 957,
        id: 442,
        incognito: false,
        left: 428,
        state: 'normal',
        tabs: [
            {
                active: true,
                audible: false,
                favIconUrl: 'http://facebook.github.io/react/favicon.ico',
                height: 862,
                highlighted: false,
                id: 443,
                incognito: false,
                index: 0,
                muted: false,
                mutedCause: '',
                pinned: false,
                selected: false,
                status: 'complete',

                // title deliberately omitted
                url: 'http://facebook.github.io/react/docs/component-api.html',
                width: 1258,
                windowId: 442,
            },
        ],
    };

    const chromeTabWindow = tabWindowUtils.makeChromeTabWindow(
        chromeWindowSnap4 as any,
    );
    const chromeTabWindowJS = JSON.parse(
        JSON.stringify(chromeTabWindow.toJS()),
    );

    const chromeTabWindowTitle = chromeTabWindow.title;

    expect(chromeTabWindow.title).toBe(chromeWindowSnap4.tabs[0].url);
});

test('attachChromeWindow', () => {
    // Let's first create the tabWindow for our saved bookmark folder:
    const tabWindow = tabWindowUtils.makeFolderTabWindow(
        testData.d3BookmarkFolder,
    );
    const tabWindowJS = JSON.parse(JSON.stringify(tabWindow.toJS()));

    /*
    console.log('makeFolderTabWindow returned:');
    console.log('>>>>>>>');
    console.log(JSON.stringify(tabWindowJS, null, 2));
    console.log('>>>>>>>');

    console.log('diffs between tabWindowJS and expected:');
    dumpDiffs(tabWindowJS, testData.d3InitialExpectedTabWindow);
    */
    // expect(tabWindowJS).toEqual(testData.d3InitialExpectedTabWindow);
    expect(tabWindowJS).toMatchSnapshot();

    expect(tabWindow.title).toBe('d3 docs');

    // Now let's do the attach:
    const updTabWindow = tabWindowUtils.updateWindow(
        tabWindow,
        testData.d3OpenedChromeWindow as any,
    );
    const updTabWindowJS = JSON.parse(JSON.stringify(updTabWindow.toJS()));

    /*
      console.log("updateTabWindow returned:")
      console.log(">>>>>>>")
      console.log(JSON.stringify(updTabWindowJS,null,2))
      console.log(">>>>>>>")

      console.log("diffs between updTabWindow (actual) and expected:")
      dumpDiffs(updTabWindowJS, testData.d3AttachedExpectedTabWindow)
  */

    expect(updTabWindowJS).toMatchSnapshot();

    const tabCount = updTabWindow.tabItems.length;
    const openCount = updTabWindow.tabItems.filter((t) => t.open).length;
    const savedCount = updTabWindow.tabItems.filter((t) => t.saved).length;

    /*
    console.log(
        'attachChromeWindow: ' +
            tabCount +
            ' total tabs, ' +
            openCount +
            ' open, ' +
            savedCount +
            ' saved'
    );
    */
    expect(tabCount).toBe(7);
    expect(openCount).toBe(6);
    expect(savedCount).toBe(6);

    // Now revert the window to saved state:
    const revTabWindow = tabWindowUtils.removeOpenWindowState(
        updTabWindow,
        false,
    );

    const revTabCount = revTabWindow.tabItems.length;
    const revOpenCount = revTabWindow.tabItems.filter((t) => t.open).length;
    const revSavedCount = revTabWindow.tabItems.filter((t) => t.saved).length;

    /*
    console.log(
        'attachChromeWindow after revert: ' +
            revTabCount +
            ' total tabs, ' +
            revOpenCount +
            ' open, ' +
            revSavedCount +
            ' saved'
    );
    */
    expect(revTabCount).toBe(6);
    expect(revOpenCount).toBe(0);
    expect(revSavedCount).toBe(6);
});

const TEST_SAVE_TAB_ID = 445;

test('saveTab', () => {
    log.setLevel(log.levels.DEBUG);
    const baseTabWindow = tabWindowUtils.makeChromeTabWindow(
        testData.chromeWindowSnap,
    );

    const tabWindowJS = JSON.parse(JSON.stringify(baseTabWindow.toJS()));

    /*
        console.log("makeChromeTabWindow returned: ")
        console.log(">>>>>")
        console.log(JSON.stringify(tabWindowJS,null,2))
        console.log(">>>>>")
    */

    const entry = baseTabWindow.findChromeTabId(TEST_SAVE_TAB_ID);

    expect(entry.length).toBe(2);
    const [tabIndex, baseTabItem] = entry;

    console.log('saveTab: base tab item: ', baseTabItem.toJS());
    expect(baseTabItem.saved).toBe(false);
    expect(baseTabItem.savedState).toBe(null);

    const saveTabNode: chrome.bookmarks.BookmarkTreeNode = {
        dateAdded: 1405471223073,
        id: '999',
        index: 5,
        parentId: '777',
        title: 'React Tutorial',
        url: 'http://facebook.github.io/react/docs/tutorial.html',
    };

    const savedTabWindow = tabWindowUtils.saveTab(
        baseTabWindow,
        baseTabItem,
        saveTabNode,
    );
    const savedTabWindowJS = JSON.parse(JSON.stringify(savedTabWindow.toJS()));

    /*
    console.log('savedTab returned: ');
    console.log('>>>>>');
    console.log(JSON.stringify(savedTabWindowJS, null, 2));
    console.log('>>>>>');
    */

    const savedEntry = savedTabWindow.findChromeTabId(TEST_SAVE_TAB_ID);

    expect(savedEntry.length).toBe(2);
    const [savedTabIndex, savedTabItem] = savedEntry;

    console.log('saveTab: savedTabItem:', savedTabItem.toJS());
    expect(savedTabItem.saved).toBe(true);
    expect(savedTabItem.savedState.bookmarkId).toBe(saveTabNode.id);
    expect(savedTabItem.savedState.bookmarkIndex).toBe(saveTabNode.index);
    expect(savedTabItem.savedState.url).toBe(saveTabNode.url);
});
