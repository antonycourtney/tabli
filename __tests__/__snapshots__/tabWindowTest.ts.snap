// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachChromeWindow 1`] = `
{
  "chromeSessionId": null,
  "expanded": null,
  "height": 0,
  "open": false,
  "openWindowId": -1,
  "saved": true,
  "savedFolderId": "431",
  "savedTitle": "d3 docs",
  "snapshot": false,
  "tabItems": [
    {
      "key": "_tabItem-562",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "432",
        "bookmarkIndex": 0,
        "title": "API Reference · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/API-Reference",
      },
    },
    {
      "key": "_tabItem-563",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "534",
        "bookmarkIndex": 1,
        "title": "D3.js - Data-Driven Documents",
        "url": "http://d3js.org/",
      },
    },
    {
      "key": "_tabItem-564",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "535",
        "bookmarkIndex": 2,
        "title": "Gallery · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Gallery",
      },
    },
    {
      "key": "_tabItem-565",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "536",
        "bookmarkIndex": 3,
        "title": "Tutorials · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Tutorials",
      },
    },
    {
      "key": "_tabItem-566",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "537",
        "bookmarkIndex": 4,
        "title": "Drag + Zoom",
        "url": "http://bl.ocks.org/mbostock/6123708",
      },
    },
    {
      "key": "_tabItem-567",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "613",
        "bookmarkIndex": 5,
        "title": "Focus+Context via Brushing",
        "url": "http://bl.ocks.org/mbostock/1667367",
      },
    },
  ],
  "width": 0,
  "windowType": "",
}
`;

exports[`attachChromeWindow 2`] = `
{
  "chromeSessionId": null,
  "expanded": null,
  "height": 0,
  "open": true,
  "openWindowId": 19,
  "saved": true,
  "savedFolderId": "431",
  "savedTitle": "d3 docs",
  "snapshot": false,
  "tabItems": [
    {
      "key": "_tabItem-580",
      "open": true,
      "openState": {
        "active": true,
        "audible": false,
        "favIconUrl": "https://assets-cdn.github.com/favicon.ico",
        "isSuspended": false,
        "muted": false,
        "openTabId": 20,
        "openTabIndex": 0,
        "pinned": false,
        "title": "API Reference · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/API-Reference",
      },
      "saved": true,
      "savedState": {
        "bookmarkId": "432",
        "bookmarkIndex": 0,
        "title": "API Reference · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/API-Reference",
      },
    },
    {
      "key": "_tabItem-581",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "isSuspended": false,
        "muted": false,
        "openTabId": 22,
        "openTabIndex": 1,
        "pinned": false,
        "title": "D3.js - Data-Driven Documents",
        "url": "https://d3js.org/",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-582",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "https://assets-cdn.github.com/favicon.ico",
        "isSuspended": false,
        "muted": false,
        "openTabId": 24,
        "openTabIndex": 2,
        "pinned": false,
        "title": "Gallery · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Gallery",
      },
      "saved": true,
      "savedState": {
        "bookmarkId": "535",
        "bookmarkIndex": 2,
        "title": "Gallery · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Gallery",
      },
    },
    {
      "key": "_tabItem-583",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "https://assets-cdn.github.com/favicon.ico",
        "isSuspended": false,
        "muted": false,
        "openTabId": 26,
        "openTabIndex": 3,
        "pinned": false,
        "title": "Tutorials · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Tutorials",
      },
      "saved": true,
      "savedState": {
        "bookmarkId": "536",
        "bookmarkIndex": 3,
        "title": "Tutorials · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Tutorials",
      },
    },
    {
      "key": "_tabItem-584",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://bl.ocks.org/favicon.png",
        "isSuspended": false,
        "muted": false,
        "openTabId": 28,
        "openTabIndex": 4,
        "pinned": false,
        "title": "Drag + Zoom - bl.ocks.org",
        "url": "http://bl.ocks.org/mbostock/6123708",
      },
      "saved": true,
      "savedState": {
        "bookmarkId": "537",
        "bookmarkIndex": 4,
        "title": "Drag + Zoom",
        "url": "http://bl.ocks.org/mbostock/6123708",
      },
    },
    {
      "key": "_tabItem-585",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://bl.ocks.org/favicon.png",
        "isSuspended": false,
        "muted": false,
        "openTabId": 30,
        "openTabIndex": 5,
        "pinned": false,
        "title": "Focus+Context via Brushing - bl.ocks.org",
        "url": "http://bl.ocks.org/mbostock/1667367",
      },
      "saved": true,
      "savedState": {
        "bookmarkId": "613",
        "bookmarkIndex": 5,
        "title": "Focus+Context via Brushing",
        "url": "http://bl.ocks.org/mbostock/1667367",
      },
    },
    {
      "key": "_tabItem-586",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "534",
        "bookmarkIndex": 1,
        "title": "D3.js - Data-Driven Documents",
        "url": "http://d3js.org/",
      },
    },
  ],
  "width": 0,
  "windowType": "normal",
}
`;

exports[`chromeTabWindow 1`] = `
{
  "chromeSessionId": null,
  "expanded": null,
  "height": 957,
  "open": true,
  "openWindowId": 442,
  "saved": false,
  "savedFolderId": "",
  "savedTitle": "",
  "snapshot": false,
  "tabItems": [
    {
      "key": "_tabItem-519",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://facebook.github.io/react/favicon.ico",
        "isSuspended": false,
        "muted": false,
        "openTabId": 443,
        "openTabIndex": 0,
        "pinned": false,
        "title": "Component API | React",
        "url": "http://facebook.github.io/react/docs/component-api.html",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-520",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://facebook.github.io/react/favicon.ico",
        "isSuspended": false,
        "muted": false,
        "openTabId": 445,
        "openTabIndex": 1,
        "pinned": false,
        "title": "Tutorial | React",
        "url": "http://facebook.github.io/react/docs/tutorial.html",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-521",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://cdn.sstatic.net/stackoverflow/img/favicon.ico?v=4f32ecc8f43d",
        "isSuspended": false,
        "muted": false,
        "openTabId": 447,
        "openTabIndex": 2,
        "pinned": false,
        "title": "javascript - Is there any proper way to integrate d3.js graphics into Facebook React application? - Stack Overflow",
        "url": "http://stackoverflow.com/questions/21903604/is-there-any-proper-way-to-integrate-d3-js-graphics-into-facebook-react-applicat",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-522",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "isSuspended": false,
        "muted": false,
        "openTabId": 449,
        "openTabIndex": 3,
        "pinned": false,
        "title": "Flux | Application Architecture for Building User Interfaces",
        "url": "http://facebook.github.io/flux/docs/overview.html#content",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-523",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://fluxxor.com/favicon.ico",
        "isSuspended": false,
        "muted": false,
        "openTabId": 451,
        "openTabIndex": 4,
        "pinned": false,
        "title": "Fluxxor - Home",
        "url": "http://fluxxor.com/",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-524",
      "open": true,
      "openState": {
        "active": false,
        "audible": false,
        "favIconUrl": "http://facebook.github.io/fixed-data-table/images/favicon-b4fca2450cb5aa407a2e106f42a92838.png",
        "isSuspended": false,
        "muted": false,
        "openTabId": 453,
        "openTabIndex": 5,
        "pinned": false,
        "title": "FixedDataTable",
        "url": "http://facebook.github.io/fixed-data-table/",
      },
      "saved": false,
      "savedState": null,
    },
    {
      "key": "_tabItem-525",
      "open": true,
      "openState": {
        "active": true,
        "audible": false,
        "favIconUrl": "https://www.google.com/images/icons/product/chrome-32.png",
        "isSuspended": false,
        "muted": false,
        "openTabId": 734,
        "openTabIndex": 6,
        "openerTabId": 449,
        "pinned": false,
        "title": "Declare Permissions - Google Chrome",
        "url": "https://developer.chrome.com/extensions/declare_permissions",
      },
      "saved": false,
      "savedState": null,
    },
  ],
  "width": 1258,
  "windowType": "normal",
}
`;

exports[`makeFolderTabWindow 1`] = `
{
  "chromeSessionId": null,
  "expanded": null,
  "height": 0,
  "open": false,
  "openWindowId": -1,
  "saved": true,
  "savedFolderId": "431",
  "savedTitle": "d3 docs",
  "snapshot": false,
  "tabItems": [
    {
      "key": "_tabItem-506",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "432",
        "bookmarkIndex": 0,
        "title": "API Reference · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/API-Reference",
      },
    },
    {
      "key": "_tabItem-507",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "534",
        "bookmarkIndex": 1,
        "title": "D3.js - Data-Driven Documents",
        "url": "http://d3js.org/",
      },
    },
    {
      "key": "_tabItem-508",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "535",
        "bookmarkIndex": 2,
        "title": "Gallery · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Gallery",
      },
    },
    {
      "key": "_tabItem-509",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "536",
        "bookmarkIndex": 3,
        "title": "Tutorials · mbostock/d3 Wiki",
        "url": "https://github.com/mbostock/d3/wiki/Tutorials",
      },
    },
    {
      "key": "_tabItem-510",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "537",
        "bookmarkIndex": 4,
        "title": "Drag + Zoom",
        "url": "http://bl.ocks.org/mbostock/6123708",
      },
    },
    {
      "key": "_tabItem-511",
      "open": false,
      "openState": null,
      "saved": true,
      "savedState": {
        "bookmarkId": "613",
        "bookmarkIndex": 5,
        "title": "Focus+Context via Brushing",
        "url": "http://bl.ocks.org/mbostock/1667367",
      },
    },
  ],
  "width": 0,
  "windowType": "",
}
`;
